pkgs/desktops/lomiri/qml/lomiri-ui-toolkit/default.nix

```diff
@@ -0,0 +1,337 @@
+      tests/unit/statesaver/tst_statesaver.cpp
+
+    # These tests try to load Suru theme icons, but override XDG_DATA_DIRS / use full paths to load them
+    substituteInPlace \
+      tests/unit/visual/tst_visual.cpp \
+      tests/unit/visual/tst_icon.{11,13}.qml \
+      tests/unit/visual/tst_imageprovider.11.qml \
+      --replace '/usr/share' '${ubuntu-themes}/share'
+
+    # lomiri-mobile rename comes from where? maybe distro-specific change, maybe they use a different Suru?
+    substituteInPlace tests/unit/visual/tst_imageprovider.11.qml \
+      --replace 'lomiri-mobile' 'ubuntu-mobile'
+
+    # Disable tests
+
+    # These tests are detected via wildcard, so delete the actual files to disable them
+    for badQmlTest in ${lib.strings.concatStringsSep " " [
+      # Requires a working Qt OpenGL context in ShapeMaterial
+      "tst_actionbar.13.qml"
+      "tst_actionlist.13.qml"
+      "tst_adaptivepagelayout.13.qml"
+      "tst_adaptivepagelayout_configuration.13.qml"
+      "tst_combobutton.11.qml"
+      "tst_combobutton.13.qml"
+      "tst_contextual_actions.13.qml"
+      "tst_focus.13.qml"
+      "tst_header.13.qml"
+      "tst_hide_chrome.11.qml"
+      "tst_listitem.12.qml"
+      "tst_listitem.13.qml"
+      "tst_listitem_actions_breaks_selectmode.12.qml"
+      "tst_listitem_extras.13.qml"
+      "tst_listitem_focus.13.qml"
+      "tst_listitem_horizontal_navigation.13.qml"
+      "tst_listitem_selectmode.12.qml"
+      "tst_listitem_selectmode.13.qml"
+      "tst_listitems_itemselector.11.qml"
+      "tst_listitems_standard.11.qml"
+      "tst_listitems_standard.13.qml"
+      "tst_lomirilistview.11.qml"
+      "tst_lomiritestcase.qml"
+      "tst_multicolumnheader.13.qml"
+      "tst_optionselector.11.qml"
+      "tst_optionselector.13.qml"
+      "tst_page_with_header.13.qml"
+      "tst_pagehead_back_action.13.qml"
+      "tst_pagehead_contents_width.13.qml"
+      "tst_pagehead_sections.13.qml"
+      "tst_pagehead_visible.13.qml"
+      "tst_pageheader.13.qml"
+      "tst_pagestack.13.qml"
+      "tst_pagestack.DEPRECATED_APPHEADER_TABS.13.qml"
+      "tst_picker.11.qml"
+      "tst_picker.13.qml"
+      "tst_popover.12.qml"
+      "tst_popover.13.qml"
+      "tst_popups_dialog.13.qml"
+      "tst_popups_pagestack.13.qml"
+      "tst_pulltorefresh_pagestack_topmargin.13.qml"
+      "tst_slider.11.qml"
+      "tst_slider.13.qml"
+      "tst_switch_bug1510919.13.qml"
+      "tst_tabs.11.qml"
+      "tst_tabs.13.qml"
+      "tst_tabs.DEPRECATED_TOOLBAR.11.qml"
+      "tst_textarea.11.qml"
+      "tst_textarea_in_flickable.11.qml"
+      "tst_textfield.11.qml"
+      "tst_textinput_common.12.qml"
+      "tst_textinput_common.13.qml"
+      "tst_toggles.13.qml"
+
+      # SignalSpy on QML shader compilers are failing for unknown reasons
+      "tst_icon.11.qml"
+      "tst_icon.13.qml"
+    ]}; do
+      find tests/unit -name $badQmlTest -delete
+    done
```

Comment:
how did you detect them? I'm guessing this was a manual effort? I can see a problem where when the package gets update some tests stay excluded even when they got fixed which would be meh. Also usually we write a comment why a particular test isn't working but I can see why you didn't do it in this case and at least for me that's not a blocker.

Reply:
Unfortunately manual yes. There are short comments describing their situations, they all have sandbox-related issues AFAICT. It's likely that one change would fix a multiple tests at the same time (at least enough to re-evaluate them), so grouping them made sense to me:

All tests in block 1 have a dependency on the `ShapeMaterial` class, which [requires a functional OpenGL context to be available](https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/976aeb5f40291044dcebb6d221821ef27bd44a78/src/LomiriToolkit/uclomirishape.cpp#L215-217) & accessible, presumably for something rendering-related (I don't remember the details, been too long). The required context is not available in the fully sandboxed build environment and `ShapeMaterial` doesn't check if requesting it failed, so later on it will segfault when it tries to use it. This exact failure (I think even with the exact same stacktrace) affects all of these tests.

All tests in block 2 attempt to observe some signal-emitting state change in the QML shader compiler via [QSignalSpies](https://doc.qt.io/qt-5/qsignalspy.html). It seems like these signal spies do not work in the build sandbox, I've also noticed this in tests of other Lomiri packages. I don't know the exact cause of this, haven't debugged it further.

Reply:
For the sake of having the exact error documented somewhere:

```
Executing dbus-test-runner --task gdb -p --quiet -n inversemousearea -m 500 -p --batch -p -ex -p 'set print thread-events off' -p -ex -p run -p -ex -p bt -p --return-child-result -p --args -p /build/source/tests/unit/inversemousearea/inversemousearea -p -maxwarnings -p 100 -p -o -p /build/source/tests/inversemousearea.xml,xunitxml -p -o -p -,txt
[ ... ]
********* Start testing of tst_InverseMouseAreaTest *********
[ ... ]
Thread 1 "inversemouseare" received signal SIGSEGV, Segmentation fault.
0x00007ffff6177746 in QOpenGLContext::functions() const () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Gui.so.5
#0  0x00007ffff6177746 in QOpenGLContext::functions() const () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Gui.so.5
#1  0x00007ffff7d1378e in ShapeMaterial::ShapeMaterial() () from /build/source/lib/libLomiriToolkit.so.5
#2  0x00007ffff7d13a6c in ShapeNode::ShapeNode() () from /build/source/lib/libLomiriToolkit.so.5
#3  0x00007ffff7d1407f in UCLomiriShape::updatePaintNode(QSGNode*, QQuickItem::UpdatePaintNodeData*) () from /build/source/lib/libLomiriToolkit.so.5
#4  0x00007ffff7273728 in QQuickWindowPrivate::updateDirtyNode(QQuickItem*) () from /nix/store/jl6ydssh76cs0bdg6wy9a9mniylacbhp-qtdeclarative-5.15.10/lib/libQt5Quick.so.5
#5  0x00007ffff7273d2b in QQuickWindowPrivate::updateDirtyNodes() () from /nix/store/jl6ydssh76cs0bdg6wy9a9mniylacbhp-qtdeclarative-5.15.10/lib/libQt5Quick.so.5
#6  0x00007ffff72758ad in QQuickWindowPrivate::syncSceneGraph() () from /nix/store/jl6ydssh76cs0bdg6wy9a9mniylacbhp-qtdeclarative-5.15.10/lib/libQt5Quick.so.5
#7  0x00007ffff722dcfb in QSGSoftwareRenderLoop::renderWindow(QQuickWindow*, bool) () from /nix/store/jl6ydssh76cs0bdg6wy9a9mniylacbhp-qtdeclarative-5.15.10/lib/libQt5Quick.so.5
#8  0x00007ffff722ec88 in QSGSoftwareRenderLoop::exposureChanged(QQuickWindow*) () from /nix/store/jl6ydssh76cs0bdg6wy9a9mniylacbhp-qtdeclarative-5.15.10/lib/libQt5Quick.so.5
#9  0x00007ffff613bd95 in QWindow::event(QEvent*) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Gui.so.5
#10 0x00007ffff697103e in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Widgets.so.5
#11 0x00007ffff54cb4b8 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Core.so.5
#12 0x00007ffff613128d in QGuiApplicationPrivate::processExposeEvent(QWindowSystemInterfacePrivate::ExposeEvent*) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Gui.so.5
#13 0x00007ffff610332c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Gui.so.5
#14 0x00007ffff1a481fa in xcbSourceDispatch(_GSource*, int (*)(void*), void*) () from /nix/store/jldi00zd8w2nwa47r779balnbwsh5ykn-qtbase-5.15.10-bin/lib/qt-5.15.10/plugins/platforms/../../../../../7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5XcbQpa.so.5
#15 0x00007ffff77e5ead in g_main_context_dispatch () from /nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4/lib/libglib-2.0.so.0
#16 0x00007ffff77e6158 in g_main_context_iterate.constprop () from /nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4/lib/libglib-2.0.so.0
#17 0x00007ffff77e61ec in g_main_context_iteration () from /nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4/lib/libglib-2.0.so.0
#18 0x00007ffff55230b6 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Core.so.5
#19 0x00007ffff54cb5f1 in QCoreApplication::processEvents(QFlags<QEventLoop::ProcessEventsFlag>, int) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Core.so.5
#20 0x00007ffff551eeb8 in QTest::qWait(int) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Core.so.5
#21 0x00000000004070b7 in QTest::mouseEvent(QTest::MouseAction, QWindow*, Qt::MouseButton, QFlags<Qt::KeyboardModifier>, QPoint, int) [clone .constprop.0] ()
#22 0x000000000040afa8 in tst_InverseMouseAreaTest::testCase_InverseMouseAreaOnTop() [clone .isra.0] ()
#23 0x00007ffff54d6a5e in QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArg>
#24 0x00007ffff7639321 in QTest::TestMethods::invokeTestOnData(int) const () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Test.so.5
#25 0x00007ffff7639ecb in QTest::TestMethods::invokeTest(int, char const*, QTest::WatchDog*) const () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Test.so.5
#26 0x00007ffff763a361 in QTest::TestMethods::invokeTests(QObject*) const () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Test.so.5
#27 0x00007ffff763a832 in QTest::qRun() () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Test.so.5
#28 0x00007ffff763abdb in QTest::qExec(QObject*, int, char**) () from /nix/store/7ibaxn4ph2dd32d5wpryvr9kphf4x9ly-qtbase-5.15.10/lib/libQt5Test.so.5
#29 0x0000000000406863 in main ()
Exited with status 65280
```

```
Executing dbus-test-runner --task gdb -p --quiet -n tst_icon.11.qml -m 500 -p --batch -p -ex -p 'set print thread-events off' -p -ex -p run -p -ex -p bt -p --return-child-result -p --args -p /build/source/tests/unit/visual/visual -p -input -p /build/source/tests/unit/visual/tst_icon.11.qml -p -maxwarnings -p 100 -p -o -p /build/source/tests/tst_icon.11.qml.xml,xunitxml -p -o -p -,txt
[ ... ]
********* Start testing of components *********
[ ... ]
FAIL!  : components::Icon::test_keyColor() 'wait for signal onStatusChanged' returned FALSE. ()
   Loc: [/build/source/tests/unit/visual/tst_icon.11.qml(125)]
```

Reply:
Is it possible to submit a patch to upstream so we can automatically disable (or add a qmake flag to disable) these broken tests in the sandbox?

Reply:
In [tests/checkresults.sh](https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/main/tests/checkresults.sh), you can add these files to `EXCEPTIONS_FILES` list,  when these tests fail, the statistic is EXCEPTIONS_FAILURES_FILES instead of FAILURES_FILES. Similarly, upstream [allows some warnings](https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/main/tests/checkresults.sh?ref_type=heads#L59-L66
) to continue to exist

A small patch should be better than deleting the file here

Reply:
I pushed a patch which adjusts `tests/checkresults.sh` for our needs. I'll still open an issue upstream to inquire if they have any ideas about the QOpenGLContext issue, and if this way to disable failing tests would be interesting enough for them to upstream.